### ISSUE TITLE
Custom method on class model, access _id via this.

### ISSUE THREAD

**Issue by wasya243** [typegoose/typegoose#815]
Hello, could you please help me? Let's say we have the following code.

```ts
import { prop, getModelForClass, Ref, DocumentType } from '@typegoose/typegoose';

class Vendor {
    _id!: mongoose.Types.ObjectId;

    @prop()
    disabled?: boolean;

    @prop({
        ref: () => Invoice,
        foreignField: 'vendorId',
        localField: '_id'
    })
    public invoices?: Ref<Invoice>[];

    public getInvoices(): Promise<DocumentType<Invoice>[]> {
        const vendorId = this._id;
        const invoiceModel = getModelForClass(Invoice);

        return invoiceModel.find({ vendorId }).exec();
    }
}

class Invoice {
    @prop({ required: true })
    created!: Date

    @prop({
        ref: () => Vendor, 
        required: true,
        foreignField: '_id',
        localField: 'vendorId',
        justOne: true
      })
      public vendor!: Ref<Vendor>;
    
      @prop({ required: true })
      public vendorId!: string;
}
```
Take a look at getInvoices method. I access _id via this in order to get all invoices by vendor. In order not to have TS error I added 
_id!: mongoose.Types.ObjectId; at the top of Vendor class. **Is it ok to do so? Or I should define _id via @prop decorator?**

**Comment by hasezoey** [comment]
> Is it ok to do so? Or I should define _id via @prop decorator?

TL;DR: yes it is ok, adding a `@prop` would overwrite all defaults

---

preferably `_id` shouldnt need to be defined, as long as you use the `DocumentType` type for `this`, like in the [Quick Start Guide: Instance Methods](https://typegoose.github.io/typegoose/docs/guides/quick-start-guide#instance-methods), also read [Why overwrite `this`](https://typegoose.github.io/typegoose/docs/guides/advanced/overwrite-this) if it is not clear on why the `this` type needs to be overwritten

PS: i dont personally recommend doing `const invoiceModel = getModelForClass(Invoice);` to get the model, because if caching is enabled this *would* work (unless a custom name is used at the `getModelForClass` function, but not here), but may also (depending on how you do models) be the call where the model actually gets created; and if caching is disabled (globally or just for that model), it would try to create a new model everytime `getModelForClass` is called, instead use the model value directly from where you export it, or use [`getModelWithString(getName(YourClass))`](https://typegoose.github.io/typegoose/docs/api/functions/get-model-with-string), which will not re-build the model (only works if caching is enabled though)

**Comment by hasezoey** [comment]
i have seen that weird type before, but i still dont know the cause, for now the workaround should be to split out the `findOne` call and the `.populate` call (they are 2 queries anyway)

btw does the type error still exist when doing the current way but removing the type-overwrite from the `.populate` call?

**Comment by wasya243** [comment]
it works without populate (just find method)

**Comment by hasezoey** [comment]
> it works without populate (just find method)

if it works without the direct `.populate` call (or the overwriting of the type?) then that means that those paths are just incompatible between the types (if it worked in typegoose 10, the maybe try to wrap your overwrite type in `DocumentType`, because in typegoose 11 `Ref` got changed to transparently use it instead of the raw class)

**Comment by wasya243** [comment]
sorry for confusion, as I said above it works with direct populate without .populate<{invoices: Invoice[]}>('invoices')

**Comment by wasya243** [comment]
works with:
```
const queriedVendor1 = await VendorModel.findOne().populate<{invoices?: DocumentType<Invoice>[]}>('invoices').exec();
const queriedVendor2 = await VendorModel.findOne().populate('invoices').exec();
```
does not work with:
```
const queriedVendor1 = await VendorModel.findOne().populate<{invoices?: Invoice[]}>('invoices').exec();
```

**Comment by wasya243** [comment]
So, basically you suggestion with wrapping with DocumentType worked.
